dangerzone/updater/signatures.py

```diff
@@ -0,0 +1,626 @@
+                "integratedTime": payload["integratedTime"],
+                "logIndex": payload["logIndex"],
+                "logID": payload["logID"],
+            },
+        },
+        "RFC3161Timestamp": sig["RFC3161Timestamp"],
+    }
+
+
+def verify_signature(signature: dict, image_digest: str, pubkey: Path) -> None:
+    """
+    Verifies that the signature matches the given public key and image digest.
+    """
+    # FIXME Also verify the identity/docker-reference field against
+    # `container_utils.expected_image_name()`
+    # e.g. ghcr.io/freedomofpress/dangerzone/dangerzone
+
+    signature_bundle = signature_to_bundle(signature)
+    try:
+        payload_bytes = b64decode(signature_bundle["Payload"])
+        payload_digest = json.loads(payload_bytes)["critical"]["image"][
+            "docker-manifest-digest"
+        ]
+    except Exception as e:
+        raise errors.SignatureVerificationError(
+            f"Unable to extract the payload digest from the signature: {e}"
+        )
+    if payload_digest != f"sha256:{image_digest}":
+        raise errors.SignatureMismatch(
+            "The given signature does not match the expected image digest "
+            f"({payload_digest}, {image_digest})"
+        )
+
+    # Note: Pass delete=False here to avoid deleting the file before usage.
+    # The file is read outside of the context manager otherwise it fails
+    # on Windows, where the file can't be opened concurrently.
+    #
+    # Using a O_TEMPORARY flag mentioned in [0] is not practical here as how
+    # cosign opens the file is not configurable.
+    #
+    # Hence the os.remove calls after calling cosign.verify_blob.
+    #
+    # [0] https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile
+    with (
+        NamedTemporaryFile(mode="w", delete=False) as signature_file,
+        NamedTemporaryFile(mode="bw", delete=False) as payload_file,
+    ):
+        json.dump(signature_bundle, signature_file)
+        signature_file.flush()
+
+        payload_file.write(payload_bytes)
+        payload_file.flush()
+
+    cosign.verify_blob(pubkey, signature_file.name, payload_file.name)
+    log.debug("Signature verified")
+
+    os.remove(signature_file.name)
+    os.remove(payload_file.name)
+
+
+@dataclass
+class Signature:
+    """Utility class to interact with signatures"""
+
+    signature: Dict
+
+    @property
+    def payload(self) -> Dict:
+        return json.loads(b64decode(self.signature["Payload"]))
+
+    @property
+    def manifest_digest(self) -> str:
+        full_digest = self.payload["critical"]["image"]["docker-manifest-digest"]
+        return full_digest.replace("sha256:", "")
+
+    @property
+    def log_index(self) -> int:
+        return self.signature["Bundle"]["Payload"]["logIndex"]
+
+
+def check_signatures_and_logindex(
+    image_str: str,
+    remote_digest: str,
+    pubkey: Path,
+    signatures: List[Dict],
+    bypass_logindex_check: bool = False,
```

Comment:
Do we really need this flag? If it's for the tests, maybe mock some calls instead?